### PR TITLE
[DROOLS-6393] Change Transformations overall management

### PR DESCRIPTION
**Pull request by gitgabrio**
@danielezonca @kostola  @jiripetrlik

See https://issues.redhat.com/browse/DROOLS-6393

Overall approach:
1) for all "transformation-related" definitions (DefineFunction, DerivedField, expressions, ecc) there is a corresponding "KiePMML" representation
2) KiePMMLTransformationDictionary/KiePMMLLocalTransformation are saved (as property) inside the generated KiePMMLModel
3) Pre/PostProcess invoke transformations over the above.

Details:

KieExpressions and KieDerivedFields expose a similar "evaluate" methods
For each "KiePMML" transformation object there is a specific code-generating "factory" that returns the code to instantiate the variable itself.

"KiePMML" transformation classes uses the "builder" pattern whenever there is an "optional" value, to simplify code-generation.

KiePMMLTransformationDictionary and KiePMMLLocalTransformation are instantiated and returned inside a specific "create" method, that is written inside the KiePMMLModel.

Other code-generation utils have been completely removed.

ExpressionFunctionUtils can not be removed because it is still in use by Scorecardmodel.
It will be addressed by https://issues.redhat.com/browse/DROOLS-6400

### CODE REVIEW

File: kie-pmml-trusty/kie-pmml-commons/src/main/java/org/kie/pmml/commons/model/KiePMMLOutputField.java [hunk 1]
```diff
@@ -64,6 +68,15 @@ public KiePMMLExpression getKiePMMLExpression() {
         return kiePMMLExpression;
     }
 
+    public Object evaluate(final List<KiePMMLDefineFunction> defineFunctions,
```
Comment (danielezonca) [inline kie-pmml-trusty/kie-pmml-commons/src/main/java/org/kie/pmml/commons/model/KiePMMLOutputField.java:71]: `@Override`

File: kie-pmml-trusty/kie-pmml-models/kie-pmml-models-regression/kie-pmml-models-regression-tests/src/test/java/org/kie/pmml/regression/tests/LinearRegressionSampleWithTransformationsTest.java [hunk 2]
```diff
@@ -117,12 +117,12 @@ public void testLogisticRegressionIrisData() {
         Assertions.assertThat(pmml4Result.getResultVariables().get(OUT_DER_CONSTANT)).isNotNull();
         Assertions.assertThat(pmml4Result.getResultVariables().get(OUT_DER_CONSTANT)).isEqualTo(CONSTANT);
         Assertions.assertThat(pmml4Result.getResultVariables().get(OUT_DER_FUN_SALARY_FIELDREF)).isNotNull();
-        Assertions.assertThat(pmml4Result.getResultVariables().get(OUT_DER_FUN_SALARY_FIELDREF)).isEqualTo(DOUBLE_CONSTANT);
+        Assertions.assertThat(pmml4Result.getResultVariables().get(OUT_DER_FUN_SALARY_FIELDREF)).isEqualTo(STRING_CONSTANT);
```
Comment (danielezonca) [inline kie-pmml-trusty/kie-pmml-models/kie-pmml-models-regression/kie-pmml-models-regression-tests/src/test/java/org/kie/pmml/regression/tests/LinearRegressionSampleWithTransformationsTest.java:120]: Is this a bugfix?

File: kie-pmml-trusty/kie-pmml-compiler/kie-pmml-compiler-commons/src/test/java/org/kie/pmml/compiler/commons/codegenfactories/KiePMMLModelFactoryUtilsTest.java [hunk 3]
```diff
@@ -267,19 +268,19 @@ public void getOutputFieldsObjectCreations() {
...
+//        final Map<String, MethodDeclaration> commonDerivedFieldsMethodMap =
+//                getDerivedFieldsMethodMap(pmmlModel.getTransformationDictionary().getDerivedFields(), arityCounter);
+//        final Map<String, MethodDeclaration> localDerivedFieldsMethodMap =
+//                getDerivedFieldsMethodMap(model.getLocalTransformations().getDerivedFields(), arityCounter);
+//        KiePMMLModelFactoryUtils.populateTransformationsInConstructor(constructorDeclaration,
+//                                                                      commonDerivedFieldsMethodMap,
+//                                                                      localDerivedFieldsMethodMap);
+//        pmmlModel.getTransformationDictionary().getDerivedFields().forEach(derivedField -> commonVerifyDerivedFieldTransformation(derivedField, commonDerivedFieldsMethodMap, "commonTransformationsMap"));
+//        model.getLocalTransformations().getDerivedFields().forEach(derivedField -> commonVerifyDerivedFieldTransformation(derivedField, localDerivedFieldsMethodMap, "localTransformationsMap"));
+//    }
```
Comment (danielezonca) [inline kie-pmml-trusty/kie-pmml-compiler/kie-pmml-compiler-commons/src/test/java/org/kie/pmml/compiler/commons/codegenfactories/KiePMMLModelFactoryUtilsTest.java]: To be removed

File: kie-pmml-trusty/kie-pmml-compiler/kie-pmml-compiler-commons/src/test/java/org/kie/pmml/compiler/commons/codegenfactories/KiePMMLModelFactoryUtilsTest.java [hunk 4]
```diff
@@ -182,10 +183,10 @@ public void addTransformationsInClassOrInterfaceDeclaration() {
...
                                                                                  model.getLocalTransformations());
-        pmmlModel.getTransformationDictionary().getDerivedFields().forEach(derivedField -> commonVerifyDerivedFieldTransformation(derivedField, null, "commonTransformationsMap"));
-        model.getLocalTransformations().getDerivedFields().forEach(derivedField -> commonVerifyDerivedFieldTransformation(derivedField, null, "localTransformationsMap"));
-        commonVerifyConstructorClass("commonTransformationsMap");
-        commonVerifyConstructorClass("localTransformationsMap");
+//        pmmlModel.getTransformationDictionary().getDerivedFields().forEach(derivedField -> commonVerifyDerivedFieldTransformation(derivedField, null, "commonTransformationsMap"));
+//        model.getLocalTransformations().getDerivedFields().forEach(derivedField -> commonVerifyDerivedFieldTransformation(derivedField, null, "localTransformationsMap"));
+//        commonVerifyConstructorClass("commonTransformationsMap");
+//        commonVerifyConstructorClass("localTransformationsMap");
```
Comment (danielezonca) [inline kie-pmml-trusty/kie-pmml-compiler/kie-pmml-compiler-commons/src/test/java/org/kie/pmml/compiler/commons/codegenfactories/KiePMMLModelFactoryUtilsTest.java]: To be removed?

File: kie-pmml-trusty/kie-pmml-commons/src/main/java/org/kie/pmml/commons/model/expressions/KiePMMLApply.java [hunk 5]
```diff
@@ -43,6 +49,36 @@ public static Builder builder(String name, List<KiePMMLExtension> extensions, St
...
+        if (kiePMMLExpressions == null) {
+            return null;
+        }
+        List<Object> expressionValues = new ArrayList<>(); // <- Insertion order matter
+        for (KiePMMLExpression kiePMMLExpression : kiePMMLExpressions) {
+            expressionValues.add(kiePMMLExpression.evaluate(defineFunctions, derivedFields, outputFields, kiePMMLNameValues));
+        }
+        BUILTIN_FUNCTIONS builtinFunction = null;
+        try {
+            builtinFunction = BUILTIN_FUNCTIONS.byName(function);
```
Comment (jiripetrlik) [inline kie-pmml-trusty/kie-pmml-commons/src/main/java/org/kie/pmml/commons/model/expressions/KiePMMLApply.java]: It seems to me like quite unusual way how to check that "function" is built in function. I think that exceptions should be used more for errors and unusual states. Would it be possible to add something like isBuiltInFunction method and use it in if statement?

**Review by jiripetrlik — COMMENTED** [review]
I have one comment here.

File: kie-pmml-trusty/kie-pmml-commons/src/main/java/org/kie/pmml/commons/model/expressions/KiePMMLApply.java (hunk 5 above)
Comment (gitgabrio) [reply kie-pmml-trusty/kie-pmml-commons/src/main/java/org/kie/pmml/commons/model/expressions/KiePMMLApply.java]: @jiripetrlik you're right, it's not nice :smile: 
Fixed. Thanks for suggestion!

File: kie-pmml-trusty/kie-pmml-commons/src/main/java/org/kie/pmml/commons/model/expressions/KiePMMLApply.java [hunk 6]
```diff
@@ -43,6 +49,31 @@ public static Builder builder(String name, List<KiePMMLExtension> extensions, St
...
+        List<Object> expressionValues = new ArrayList<>(); // <- Insertion order matter
+        for (KiePMMLExpression kiePMMLExpression : kiePMMLExpressions) {
+            expressionValues.add(kiePMMLExpression.evaluate(defineFunctions, derivedFields, outputFields, kiePMMLNameValues));
+        }
+        if (BUILTIN_FUNCTIONS.isBUILTIN_FUNCTIONS(function)) {
+            BUILTIN_FUNCTIONS builtinFunction = BUILTIN_FUNCTIONS.byName(function);
+            return builtinFunction.getValue(expressionValues.toArray(new Object[0]));
+        } else {
+            final KiePMMLDefineFunction definedFunction = defineFunctions
+                    .parallelStream()
```
Comment (danielezonca) [inline kie-pmml-trusty/kie-pmml-commons/src/main/java/org/kie/pmml/commons/model/expressions/KiePMMLApply.java]: Why parallel stream? I think we should avoid parallel stream

File: kie-pmml-trusty/kie-pmml-commons/src/main/java/org/kie/pmml/commons/model/expressions/KiePMMLFieldRef.java [hunk 7]
```diff
@@ -63,4 +84,35 @@ public boolean equals(Object o) {
     public int hashCode() {
         return Objects.hash(super.hashCode(), mapMissingTo);
     }
+
+    private Optional<Object> getFromKiePMMLNameValues(final List<KiePMMLNameValue> kiePMMLNameValues) {
+        return kiePMMLNameValues
+                .parallelStream()
```
Comment (danielezonca) [inline kie-pmml-trusty/kie-pmml-commons/src/main/java/org/kie/pmml/commons/model/expressions/KiePMMLFieldRef.java]: Same as above

File: kie-pmml-trusty/kie-pmml-commons/src/main/java/org/kie/pmml/commons/model/expressions/KiePMMLFieldRef.java [hunk 8]
```diff
@@ -63,4 +84,35 @@ public boolean equals(Object o) {
...
+                .findFirst()
+                .map(KiePMMLNameValue::getValue);
+    }
+
+    private Optional<Object> getFromDerivedFields(final List<KiePMMLDefineFunction> defineFunctions,
+                                                  final List<KiePMMLDerivedField> derivedFields,
+                                                  final List<KiePMMLOutputField> outputFields,
+                                                  final List<KiePMMLNameValue> kiePMMLNameValues) {
+        return derivedFields
+                .parallelStream()
```
Comment (danielezonca) [inline kie-pmml-trusty/kie-pmml-commons/src/main/java/org/kie/pmml/commons/model/expressions/KiePMMLFieldRef.java]: Same here

File: kie-pmml-trusty/kie-pmml-commons/src/main/java/org/kie/pmml/commons/model/expressions/KiePMMLFieldRef.java [hunk 9]
```diff
@@ -63,4 +84,35 @@ public boolean equals(Object o) {
...
+                .map(derivedField -> derivedField.evaluate(defineFunctions, derivedFields, outputFields,
+                                                           kiePMMLNameValues));
+    }
+
+    private Optional<Object> getFromOutputFields(final List<KiePMMLDefineFunction> defineFunctions,
+                                                 final List<KiePMMLDerivedField> derivedFields,
+                                                 final List<KiePMMLOutputField> outputFields,
+                                                 final List<KiePMMLNameValue> kiePMMLNameValues) {
+        return outputFields
+                .parallelStream()
```
Comment (danielezonca) [inline kie-pmml-trusty/kie-pmml-commons/src/main/java/org/kie/pmml/commons/model/expressions/KiePMMLFieldRef.java]: Same here

File: kie-pmml-trusty/kie-pmml-commons/src/main/java/org/kie/pmml/commons/transformations/KiePMMLParameterField.java [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+    }
+
+    public String getDisplayName() {
+        return displayName;
+    }
+
+    public static class Builder extends AbstractKiePMMLComponent.Builder<KiePMMLParameterField> {
+
+        private Builder(String name, List<KiePMMLExtension> extensions) {
+            super("ParameterField-", () -> new KiePMMLParameterField(name, extensions));
```
Comment (danielezonca) [inline kie-pmml-trusty/kie-pmml-commons/src/main/java/org/kie/pmml/commons/transformations/KiePMMLParameterField.java:60]: Is `ParameterField-` a "keyword"/naming convention used somewhere else? Should we move it to constant?

File: kie-pmml-trusty/kie-pmml-commons/src/test/java/org/kie/pmml/commons/model/expressions/KiePMMLConstantTest.java [hunk 11]
```diff
@@ -0,0 +1,34 @@
...
+ * limitations under the License.
+ */
+
+package org.kie.pmml.commons.model.expressions;
+
+import java.util.Collections;
+
+import org.junit.Test;
+
+import static org.junit.Assert.*;
```
Comment (danielezonca) [inline kie-pmml-trusty/kie-pmml-commons/src/test/java/org/kie/pmml/commons/model/expressions/KiePMMLConstantTest.java]: Wildcard import

File: kie-pmml-trusty/kie-pmml-commons/src/test/java/org/kie/pmml/commons/model/expressions/KiePMMLFieldRefTest.java [hunk 12]
```diff
@@ -0,0 +1,101 @@
...
+
+import org.junit.BeforeClass;
+import org.junit.Test;
+import org.kie.api.pmml.ParameterInfo;
+import org.kie.pmml.api.enums.DATA_TYPE;
+import org.kie.pmml.api.enums.OP_TYPE;
+import org.kie.pmml.commons.model.tuples.KiePMMLNameValue;
+import org.kie.pmml.commons.transformations.KiePMMLDerivedField;
+
+import static org.junit.Assert.*;
```
Comment (danielezonca) [inline kie-pmml-trusty/kie-pmml-commons/src/test/java/org/kie/pmml/commons/model/expressions/KiePMMLFieldRefTest.java]: Wildcard import

**Comment by gitgabrio** [comment]
Jenkins retest this

File: kie-pmml-trusty/kie-pmml-commons/src/test/java/org/kie/pmml/commons/transformations/KiePMMLDefineFunctionTest.java [hunk 13]
```diff
@@ -0,0 +1,116 @@
...
+import java.util.Arrays;
+import java.util.Collections;
+
+import org.junit.Test;
+import org.kie.pmml.api.enums.OP_TYPE;
+import org.kie.pmml.commons.model.expressions.KiePMMLApply;
+import org.kie.pmml.commons.model.expressions.KiePMMLConstant;
+import org.kie.pmml.commons.model.expressions.KiePMMLFieldRef;
+
+import static org.junit.Assert.*;
```
Comment (danielezonca) [inline kie-pmml-trusty/kie-pmml-commons/src/test/java/org/kie/pmml/commons/transformations/KiePMMLDefineFunctionTest.java]: Wildcard import

File: kie-pmml-trusty/kie-pmml-compiler/kie-pmml-compiler-commons/src/test/java/org/kie/pmml/compiler/commons/testutils/CodegenTestUtils.java [hunk 14]
```diff
@@ -72,6 +72,28 @@ public static void commonValidateCompilation(BlockStmt body, List<Parameter> par
         }
     }
 
+    public static void commonValidateCompilationWithImports(BlockStmt body, List<Class<?>> imports) {
```
Comment (danielezonca) [inline kie-pmml-trusty/kie-pmml-compiler/kie-pmml-compiler-commons/src/test/java/org/kie/pmml/compiler/commons/testutils/CodegenTestUtils.java:75]: Is there any other way to perform this check that doesn't require to compile the class? If not what about mention in method name that there is a compilation inside?

File: kie-pmml-trusty/kie-pmml-compiler/kie-pmml-compiler-commons/src/main/java/org/kie/pmml/compiler/commons/codegenfactories/KiePMMLExpressionFactory.java [hunk 15]
```diff
@@ -0,0 +1,65 @@
...
+ */
+public class KiePMMLExpressionFactory {
+
+    private static final String EXPRESSION_NOT_MANAGED = "Expression %s not managed";
+
+    private KiePMMLExpressionFactory() {
+        // Avoid instantiation
+    }
+
+    public static BlockStmt getKiePMMLExpression(final String variableName, final org.dmg.pmml.Expression expression) {
```
Comment (danielezonca) [inline kie-pmml-trusty/kie-pmml-compiler/kie-pmml-compiler-commons/src/main/java/org/kie/pmml/compiler/commons/codegenfactories/KiePMMLExpressionFactory.java:38]: Please review this method it contains many commented code

**Review by jiripetrlik — APPROVED** [review]
@gitgabrio Thank you for the changes. LGTM

File: kie-pmml-trusty/kie-pmml-commons/src/main/java/org/kie/pmml/commons/model/KiePMMLOutputField.java (hunk 1 above)
Comment (gitgabrio) [reply kie-pmml-trusty/kie-pmml-commons/src/main/java/org/kie/pmml/commons/model/KiePMMLOutputField.java:71]: `KiePMMLOutputField` does not inherit `evaluate`

File: kie-pmml-trusty/kie-pmml-models/kie-pmml-models-regression/kie-pmml-models-regression-tests/src/test/java/org/kie/pmml/regression/tests/LinearRegressionSampleWithTransformationsTest.java (hunk 2 above)
Comment (gitgabrio) [reply kie-pmml-trusty/kie-pmml-models/kie-pmml-models-regression/kie-pmml-models-regression-tests/src/test/java/org/kie/pmml/regression/tests/LinearRegressionSampleWithTransformationsTest.java:120]: yes - previously the constant (that in the pmml is actually a string) was wrongly interpreted as double

File: kie-pmml-trusty/kie-pmml-compiler/kie-pmml-compiler-commons/src/test/java/org/kie/pmml/compiler/commons/codegenfactories/KiePMMLModelFactoryUtilsTest.java (hunk 3 above)
Comment (gitgabrio) [reply kie-pmml-trusty/kie-pmml-compiler/kie-pmml-compiler-commons/src/test/java/org/kie/pmml/compiler/commons/codegenfactories/KiePMMLModelFactoryUtilsTest.java]: @danielezonca  done

File: kie-pmml-trusty/kie-pmml-compiler/kie-pmml-compiler-commons/src/test/java/org/kie/pmml/compiler/commons/codegenfactories/KiePMMLModelFactoryUtilsTest.java (hunk 4 above)
Comment (gitgabrio) [reply kie-pmml-trusty/kie-pmml-compiler/kie-pmml-compiler-commons/src/test/java/org/kie/pmml/compiler/commons/codegenfactories/KiePMMLModelFactoryUtilsTest.java]: @danielezonca done

File: kie-pmml-trusty/kie-pmml-commons/src/main/java/org/kie/pmml/commons/model/expressions/KiePMMLApply.java (hunk 6 above)
Comment (kostola) [reply kie-pmml-trusty/kie-pmml-commons/src/main/java/org/kie/pmml/commons/model/expressions/KiePMMLApply.java]: +1

File: kie-pmml-trusty/kie-pmml-commons/src/main/java/org/kie/pmml/commons/model/expressions/KiePMMLFieldRef.java (hunk 7 above)
Comment (kostola) [reply kie-pmml-trusty/kie-pmml-commons/src/main/java/org/kie/pmml/commons/model/expressions/KiePMMLFieldRef.java]: +1

File: kie-pmml-trusty/kie-pmml-commons/src/main/java/org/kie/pmml/commons/model/expressions/KiePMMLFieldRef.java (hunk 8 above)
Comment (kostola) [reply kie-pmml-trusty/kie-pmml-commons/src/main/java/org/kie/pmml/commons/model/expressions/KiePMMLFieldRef.java]: +1

File: kie-pmml-trusty/kie-pmml-commons/src/main/java/org/kie/pmml/commons/model/expressions/KiePMMLFieldRef.java (hunk 9 above)
Comment (kostola) [reply kie-pmml-trusty/kie-pmml-commons/src/main/java/org/kie/pmml/commons/model/expressions/KiePMMLFieldRef.java]: +1

File: kie-pmml-trusty/kie-pmml-commons/src/main/java/org/kie/pmml/commons/transformations/KiePMMLParameterField.java (hunk 10 above)
Comment (gitgabrio) [reply kie-pmml-trusty/kie-pmml-commons/src/main/java/org/kie/pmml/commons/transformations/KiePMMLParameterField.java:60]: @danielezonca nope. All those builders create an "identifier" whose name starts with the original element (e.g. ParameterField, TruePRedicate, etc etc)

File: kie-pmml-trusty/kie-pmml-commons/src/test/java/org/kie/pmml/commons/model/expressions/KiePMMLConstantTest.java (hunk 11 above)
Comment (gitgabrio) [reply kie-pmml-trusty/kie-pmml-commons/src/test/java/org/kie/pmml/commons/model/expressions/KiePMMLConstantTest.java]: @danielezonca done

File: kie-pmml-trusty/kie-pmml-commons/src/main/java/org/kie/pmml/commons/model/expressions/KiePMMLApply.java (hunk 6 above)
Comment (gitgabrio) [reply kie-pmml-trusty/kie-pmml-commons/src/main/java/org/kie/pmml/commons/model/expressions/KiePMMLApply.java]: @danielezonca done

File: kie-pmml-trusty/kie-pmml-commons/src/main/java/org/kie/pmml/commons/model/expressions/KiePMMLFieldRef.java (hunk 7 above)
Comment (gitgabrio) [reply kie-pmml-trusty/kie-pmml-commons/src/main/java/org/kie/pmml/commons/model/expressions/KiePMMLFieldRef.java]: @danielezonca done

File: kie-pmml-trusty/kie-pmml-commons/src/main/java/org/kie/pmml/commons/model/expressions/KiePMMLFieldRef.java (hunk 8 above)
Comment (gitgabrio) [reply kie-pmml-trusty/kie-pmml-commons/src/main/java/org/kie/pmml/commons/model/expressions/KiePMMLFieldRef.java]: @danielezonca done

File: kie-pmml-trusty/kie-pmml-commons/src/main/java/org/kie/pmml/commons/model/expressions/KiePMMLFieldRef.java (hunk 9 above)
Comment (gitgabrio) [reply kie-pmml-trusty/kie-pmml-commons/src/main/java/org/kie/pmml/commons/model/expressions/KiePMMLFieldRef.java]: @danielezonca done

File: kie-pmml-trusty/kie-pmml-commons/src/test/java/org/kie/pmml/commons/model/expressions/KiePMMLFieldRefTest.java (hunk 12 above)
Comment (gitgabrio) [reply kie-pmml-trusty/kie-pmml-commons/src/test/java/org/kie/pmml/commons/model/expressions/KiePMMLFieldRefTest.java]: @danielezonca done

File: kie-pmml-trusty/kie-pmml-commons/src/test/java/org/kie/pmml/commons/transformations/KiePMMLDefineFunctionTest.java (hunk 13 above)
Comment (gitgabrio) [reply kie-pmml-trusty/kie-pmml-commons/src/test/java/org/kie/pmml/commons/transformations/KiePMMLDefineFunctionTest.java]: @danielezonca done

File: kie-pmml-trusty/kie-pmml-compiler/kie-pmml-compiler-commons/src/test/java/org/kie/pmml/compiler/commons/testutils/CodegenTestUtils.java (hunk 14 above)
Comment (gitgabrio) [reply kie-pmml-trusty/kie-pmml-compiler/kie-pmml-compiler-commons/src/test/java/org/kie/pmml/compiler/commons/testutils/CodegenTestUtils.java:75]: @danielezonca I exactly wanted to do compilation to verify correctness of generated code. 
I could rename those methods to "compilation" or similar, but TBH I though `commonValidateCompilation(...)` was clear enough....

File: kie-pmml-trusty/kie-pmml-compiler/kie-pmml-compiler-commons/src/main/java/org/kie/pmml/compiler/commons/codegenfactories/KiePMMLExpressionFactory.java (hunk 15 above)
Comment (gitgabrio) [reply kie-pmml-trusty/kie-pmml-compiler/kie-pmml-compiler-commons/src/main/java/org/kie/pmml/compiler/commons/codegenfactories/KiePMMLExpressionFactory.java:38]: @danielezonca done

**Review by danielezonca — APPROVED** [review]
LGTM 👍

**Pull request merged**